docs/plots/backends.rst

```diff
@@ -0,0 +1,389 @@
+but they are allowed to have return values if that is helpful for their internal
+implementation, e.g. in the `~sherpa.plot.pylab_backend.PylabBackend` backend, plotting a line might return a
+line object so that error bars plotted later can use ``line.color`` to match the
+color of that line.
+
+Creating plots and panels, clearing and overplotting
+----------------------------------------------------
+.. todo::
+   Add details
+
+Each of the plotting functions above accepts
+the following arguments: title, xlabel, ylabel, xlog, ylog, overplot, clearwindow
+
+Multi-panels plot can be set with ``set_subplot`` and ``set_jointplot``.
+See `sherpa.plot.backends.BaseBackend` for a listing of the calling signature.
+
+
+Interaction with interactive plots in the UI
+--------------------------------------------
+Each backend also acts as a context manager: All plotting commands
+in the UI are wrapped in a ``with`` statement like this::
+
+    >>> with sherpa.backend():
+    ...     plotobj.plot()
+
+That allows the backend to execute any finishing code after the plots are done,
+e.g. to save the plot to disk or to display it in a window.
+
+Other methods
+--------------
+
+Backends need to have a few more methods:
+
+- ``as_html_XXX`` (where XXX is a plot type) that are used for interactive
+  display in the notebook with ``_repr_html_``.  These functions take a plot
+  object and return a html representation as a string.
+- ``get_XXX_plot/hist_prefs`` (where XXX is a plot type) which returns a
+  dictionary of preferences that is used for displaying this plot.
+- `~sherpa.plot.backend.BasicBackend.get_latex_for_string` to format latex in strings.
+- ``colorlist(n)`` generates a list of n distinct colors. In backends that only have a
+  limited number of colors available, the list might repeat.
+
+
+How do I implement a new backend?
+=================================
+
+Example for a modified backend
+------------------------------
+Since Sherpa backends are defined using inheritance, new backends can be
+created to change the appearance of plots in ways that are beyond the
+options offered as keyword arguments for the existing backends. As an example,
+the `~sherpa.plot.pylab_backend.PylabErrorArea` backend changes the way that
+x-errors are visualized in plots from simple error bars to a shaded area.
+
+.. literalinclude:: ../../sherpa/plot/pylab_area_backend.py
+   :language: python
+   :lines: 38-99
+
+.. todo::
+
+   Put code example in here that uses it?
+
+Testing backends and plotting code
+----------------------------------
+Currently, Sherpa does not employ pixel-level tests that compare a generated
+image pixel-by-pixel to a reference image. While testing like that guarantees
+that any and all changes are found, they are susceptible to failing for
+reasons unrelated to Sherpa, such as minor changes in the default of
+:term:`matplotlib`.
+
+Instead, the plotting tests in Sherpa fall into the following categories:
```

Comment:
Not for this PR, and I don't want to derail this dicussion so I'll mention it and then move on, but I wonder if we should move backend-specific tests into their own test file. Then all the tests could be easily gated for the presence of a backend, and make it easier to see what checks we make for a particular backend.

There are many downsides, including

- sometimes you want to check with different backends that may have some slightly-different test/code but is testing the same thing
- you have to distinguish between generic "just check we work with a backend" vs "test out this particular behavior with a given backend"
- we have a lot of tests scattered around and cleaning this up is a lot of work for little obvious gain

Reply:
There are a few backend specific tests (see e.g. `test_pylab_specific.py`) but most tests that we have apply 90% of the test code to all text backends. So, yes, we need a better way to account for the differences, but we don't want to have multiple big files files that have nearly identical tests in them.